qiime2/sdk/action.py

```diff
@@ -14,30 +14,62 @@
 
 import decorator
 import dill
+import parsl
+from parsl.app.app import python_app, join_app
 
 import qiime2.sdk
 import qiime2.core.type as qtype
 import qiime2.core.archive as archive
 from qiime2.core.util import LateBindingAttribute, DropFirstParameter, tuplize
+from qiime2.sdk.parsl_config import PARSL_CONFIG, setup_parsl
+from qiime2.sdk.proxy import Proxy
 
 
-def _subprocess_apply(action, ctx, args, kwargs):
+def _subprocess_apply(action, ctx, execution_ctx, args, kwargs):
     # We with in the cache here to make sure archiver.load* puts things in the
     # right cache
     with ctx.cache:
-        exe = action._bind(lambda: qiime2.sdk.Context(parent=ctx))
+        exe = action._bind(
+            lambda: qiime2.sdk.Context(parent=ctx), execution_ctx)
         results = exe(*args, **kwargs)
 
         return results
 
 
+def run_parsl_action(action, ctx, execution_ctx, args, kwargs, inputs=[]):
+    """This is what the parsl app itself actually runs. It's basically just a
+    wrapper around our QIIME 2 action
+    """
+    remapped_kwargs = {}
+    for key, value in kwargs.items():
+        if isinstance(value, Proxy):
+            remapped_kwargs[key] = value.get_element(inputs[value._future_])
```

Comment:
We might want to use a trivial variable here to communicate what we're working with, for example:
```
resolved_results = inputs[value._future_] # should this be _future_index_ instead?
... = value.get_element(resolved_results)
```